vault/core_metrics.go

```diff
@@ -0,0 +1,112 @@
+package vault
+
+import (
+	"context"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/vault/helper/metricsutil"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+// TODO: move emitMetrics into this file.
+
+type kvMount struct {
+	Namespace  *namespace.Namespace
+	MountPoint string
+	Version    string
+	NumSecrets int
+}
+
+func (c *Core) findKvMounts() []*kvMount {
+	mounts := make([]*kvMount, 0)
+
+	c.mountsLock.RLock()
+	defer c.mountsLock.RUnlock()
+
+	for _, entry := range c.mounts.Entries {
+		if entry.Type == "kv" {
+			version, ok := entry.Options["version"]
+			if !ok {
+				version = "1"
+			}
+			mounts = append(mounts, &kvMount{
+				Namespace:  entry.namespace,
+				MountPoint: entry.Path,
+				Version:    version,
+				NumSecrets: 0,
+			})
+		}
+	}
+	return mounts
+}
+
+func (c *Core) kvSecretGaugeCollector(ctx context.Context) ([]metricsutil.GaugeLabelValues, error) {
+	// Find all KV mounts
+	mounts := c.findKvMounts()
+	results := make([]metricsutil.GaugeLabelValues, len(mounts))
+
+	// Context must have root namespace
+	ctx = namespace.RootContext(ctx)
+
+	// Route list requests to all the identified mounts.
+	// (All of these will show up as activity in the vault.route metric.)
+	// Then we have to explore each subdirectory.
+	for i, m := range mounts {
+		results[i].Labels = []metrics.Label{
+			metricsutil.NamespaceLabel(m.Namespace),
+			{"mount_point", m.MountPoint},
+		}
+
+		var subdirectories []string
+		if m.Version == "1" {
+			subdirectories = []string{m.MountPoint}
+		} else {
+			subdirectories = []string{m.MountPoint + "metadata/"}
+		}
+
+		for len(subdirectories) > 0 {
+			// If collection was cancelled, return an empty array.
+			select {
+			case <-ctx.Done():
+				return []metricsutil.GaugeLabelValues{}, nil
+			default:
+				break
+			}
+
+			currentDirectory := subdirectories[0]
+			subdirectories = subdirectories[1:]
+
+			listRequest := &logical.Request{
+				Operation: logical.ListOperation,
+				Path:      currentDirectory,
+			}
+			resp, err := c.router.Route(ctx, listRequest)
+			if err != nil {
+				c.logger.Error("failed to perform internal KV list", "mount_point", m.MountPoint, "error", err)
+				// TODO: mark just one gauge as failed?
```

Comment:
This could happen in a healthy system if something gets unmounted between when we list the mounts and run this code, right?  In which case it's probably nicer for us to continue with the other mounts.

Reply:
That's a good point, I wasn't clear on whether it would return an error.  (One of my frustrations with the list operation is how often it just returns nothing when it *should* return an error.)

We probably need to report the metrics.collection.error count within this code to get coverage that the counts might be bad, if we don't report a top-level error. I prefer having the text message for troubleshooting, though, and that seems inappropriate to put in a metric label.

It would be nice to understand expected errors like "that mount doesn't exist any more", I will see if I can make the error handling a little smarter.

Reply:
Improved with the new handling of errors (by increasing the error counter per mount that fails.)